File: LifeTrackerFullStack/LifeTrackerBackend/Models/Hero.cs
namespace LifeTracker.Models;

public class Hero
{
    public int Id { get; set; }
    public string Name { get; set; } = string.Empty;
    public int Level { get; set; } = 1;
    public int XP { get; set; } = 0;
    public int MaxXP { get; set; } = 100;
    public int HP { get; set; } = 100;

    public void GainXP(int amount)
    {
        XP += amount;
        
        while (XP >= MaxXP)
        {
            XP -= MaxXP;
            Level++;
            MaxXP = Level * 100;
        }
    }
}

File: LifeTrackerMobile/app/src/main/java/com/lifetracker/mobile/data/Hero.kt
package com.lifetracker.mobile.data

import androidx.room.Entity
import androidx.room.PrimaryKey
import com.google.gson.annotations.SerializedName

@Entity(tableName = "heroes")
data class Hero(
    @PrimaryKey
    val id: Int,

    @SerializedName("name")
    val name: String,

    @SerializedName("level")
    val level: Int,

    @SerializedName("xp")
    val xp: Int,

    @SerializedName("maxXP")
    val experienceToNextLevel: Int,

    @SerializedName("hp")
    val hp: Int
) {
    fun getExperienceProgress(): Float {
        return if (experienceToNextLevel > 0) {
            xp.toFloat() / experienceToNextLevel.toFloat()
        } else {
            0f
        }
    }
}


